File: swift-bridge/Sources/SpriteKitBridge/Texture.swift
import CoreGraphics
import SpriteKit

@_cdecl("sk_texture_image_named")
public func sk_texture_image_named(_ name: UnsafePointer<CChar>?) -> UnsafeMutableRawPointer? {
    guard let name else { return nil }
    return skRetain(SKTexture(imageNamed: String(cString: name)))
}

@_cdecl("sk_texture_from_cg_image")
public func sk_texture_from_cg_image(_ imageHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let image = skBorrowCGImage(imageHandle) else { return nil }
    return skRetain(SKTexture(cgImage: image))
}

@_cdecl("sk_texture_from_rgba_bytes")
public func sk_texture_from_rgba_bytes(
    _ bytes: UnsafeRawPointer?,
    _ length: Int,
    _ width: Int,
    _ height: Int
) -> UnsafeMutableRawPointer? {
    guard let bytes else { return nil }
    let data = Data(bytes: bytes, count: length)
    return skRetain(SKTexture(data: data, size: CGSize(width: width, height: height)))
}

@_cdecl("sk_texture_subrect")
public func sk_texture_subrect(
    _ textureHandle: UnsafeMutableRawPointer?,
    _ x: Double, _ y: Double, _ width: Double, _ height: Double
) -> UnsafeMutableRawPointer? {
    guard let texture: SKTexture = skBorrow(textureHandle) else { return nil }
    let rect = CGRect(x: x, y: y, width: width, height: height)
    return skRetain(SKTexture(rect: rect, in: texture))
}

@_cdecl("sk_texture_get_size_w")
public func sk_texture_get_size_w(_ textureHandle: UnsafeMutableRawPointer?) -> Double {
    guard let texture: SKTexture = skBorrow(textureHandle) else { return 0 }
    return Double(texture.size().width)
}

@_cdecl("sk_texture_get_size_h")
public func sk_texture_get_size_h(_ textureHandle: UnsafeMutableRawPointer?) -> Double {
    guard let texture: SKTexture = skBorrow(textureHandle) else { return 0 }
    return Double(texture.size().height)
}

@_cdecl("sk_texture_get_filtering_mode")
public func sk_texture_get_filtering_mode(_ textureHandle: UnsafeMutableRawPointer?) -> Int32 {
    guard let texture: SKTexture = skBorrow(textureHandle) else { return 0 }
    return Int32(texture.filteringMode.rawValue)
}

@_cdecl("sk_texture_set_filtering_mode")
public func sk_texture_set_filtering_mode(_ textureHandle: UnsafeMutableRawPointer?, _ mode: Int32) {
    guard let texture: SKTexture = skBorrow(textureHandle) else { return }
    texture.filteringMode = SKTextureFilteringMode(rawValue: Int(mode)) ?? .linear
}

@_cdecl("sk_texture_get_uses_mipmaps")
public func sk_texture_get_uses_mipmaps(_ textureHandle: UnsafeMutableRawPointer?) -> Bool {
    guard let texture: SKTexture = skBorrow(textureHandle) else { return false }
    return texture.usesMipmaps
}

@_cdecl("sk_texture_set_uses_mipmaps")
public func sk_texture_set_uses_mipmaps(_ textureHandle: UnsafeMutableRawPointer?, _ uses: Bool) {
    guard let texture: SKTexture = skBorrow(textureHandle) else { return }
    texture.usesMipmaps = uses
}

File: src/texture.rs
use apple_cf::cg::{CGRect, CGSize};

use crate::ffi;
use crate::private::{cstring_from_str, handle_type};

handle_type!(Texture);

/// `SKTexture` filtering mode.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(i32)]
pub enum TextureFilteringMode {
    Nearest = 0,
    Linear = 1,
}

impl TextureFilteringMode {
    #[must_use]
    pub const fn from_raw(value: i32) -> Self {
        match value {
            0 => Self::Nearest,
            _ => Self::Linear,
        }
    }
}

impl Texture {
    /// Loads a texture by asset-catalogue / bundle name.
    #[must_use]
    pub fn image_named(name: &str) -> Option<Self> {
        let name = cstring_from_str(name)?;
        unsafe { Self::from_raw(ffi::sk_texture_image_named(name.as_ptr())) }
    }

    /// Creates a texture from a `CGImage` (takes a borrowed raw pointer from `apple-cf`).
    ///
    /// # Safety
    ///
    /// `image_ptr` must be a valid, live `CGImage` pointer.
    #[must_use]
    pub unsafe fn from_cg_image_ptr(image_ptr: *mut core::ffi::c_void) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_texture_from_cg_image(image_ptr)) }
    }

    /// Creates a texture from raw RGBA bytes.
    ///
    /// `bytes` must have length `width * height * 4`.
    #[must_use]
    pub fn from_rgba_bytes(bytes: &[u8], width: usize, height: usize) -> Option<Self> {
        if bytes.len() < width.saturating_mul(height).saturating_mul(4) {
            return None;
        }
        unsafe {
            Self::from_raw(ffi::sk_texture_from_rgba_bytes(
                bytes.as_ptr(),
                bytes.len(),
                width,
                height,
            ))
        }
    }

    /// Returns a sub-rectangle texture from this texture.
    ///
    /// `rect` coordinates are in normalised (0..1) UV space.
    #[must_use]
    pub fn subrect(&self, rect: CGRect) -> Option<Self> {
        unsafe {
            Self::from_raw(ffi::sk_texture_subrect(
                self.ptr,
                rect.x,
                rect.y,
                rect.width,
                rect.height,
            ))
        }
    }

    #[must_use]
    pub fn size(&self) -> CGSize {
        let w = unsafe { ffi::sk_texture_get_size_w(self.ptr) };
        let h = unsafe { ffi::sk_texture_get_size_h(self.ptr) };
        CGSize::new(w, h)
    }

    #[must_use]
    pub fn filtering_mode(&self) -> TextureFilteringMode {
        TextureFilteringMode::from_raw(unsafe { ffi::sk_texture_get_filtering_mode(self.ptr) })
    }

    pub fn set_filtering_mode(&self, mode: TextureFilteringMode) {
        unsafe { ffi::sk_texture_set_filtering_mode(self.ptr, mode as i32) };
    }

    #[must_use]
    pub fn uses_mipmaps(&self) -> bool {
        unsafe { ffi::sk_texture_get_uses_mipmaps(self.ptr) }
    }

    pub fn set_uses_mipmaps(&self, uses: bool) {
        unsafe { ffi::sk_texture_set_uses_mipmaps(self.ptr, uses) };
    }
}
